### PR TITLE
Fix encoding error while uploading segmentations

### PR DESCRIPTION
**Pull request by dvischi**

### CODE REVIEW

File: src/python/tmclient/api.py [hunk 1]
```diff
@@ -1647,7 +1647,7 @@ def _upload_segmentation_image(self, mapobject_type_name,
         )
         npz_file = BytesIO()
         np.savez_compressed(npz_file, segmentation=image)
-        npz_file_str = base64.b64encode(npz_file.getvalue())
+        npz_file_str = base64.encodestring(npz_file.getvalue())
```
[the remark this answers is not in the document]
Comment (dvischi) [reply src/python/tmclient/api.py]: Agree - I reset the deprecated function call. However, given a byte stream input Python3's base64 encoding returns a byte stream as output while Python2 returns a string.
I added a switch for Python3 to convert the byte stream into a UTF-8 string which can then be serialized while transferring data to the TmServer.
If anyone has another suggestions, please let me know.

**Comment by riccardomurri** [comment]
I had a look at how the "bytes" issue should be handled, as agreed in today's meeting.

First of all: the issue with not being able to serialize Python *bytestrings* is not an issue with the "requests" library, rather with the JSON encoding, as the following snippet shows:

```
 $ python3
...
>>> import json
>>> x = b'abcdef'
>>> json.dumps(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
TypeError: Object of type 'bytes' is not JSON serializable
```

So the errors happens at [line 1667][a] but it's raised by the JSON encoder, not by requests "POST" action.

We could have [request upload binary data as a file/stream][b] but that would require changing the API server-side to accept a file upload instead of the NumPy array encoded as a string as it does now.  I think this is the approach we should take longer-term; exactly because it allows us to just upload a NumPy array dump without worrying about conversion to string, etc.  Shorter-term, I think your fix is OK but I would like to rephrase it a bit:

- add a `# FIXME` comment stating that this is a workaround and maybe reference this discussion

- I would test for the needed feature that for the Python version;
  since the issue is that Python3's "bytestrings" are not "strings" I
  would use this code instead::

```
npz_file_str = base64.encodestring(npz_file.getvalue())
if bytes != str:
    # turn it into a `str` object on Python 3.x,
    # otherwise JSON encoding below fails
    npz_file_str = npz_file_str.decode()
```

- there's still an optimization that could be done in Python 3.x, using `npz_file.getbuffer()` instead of `npz_file.getvalue()` (avoid making a copy of the memory buffer) -- but I guess this is out of scope of this PR ;-)

[a]: https://github.com/dvischi/TmClient/blob/a5fc5d8a8ce1ef0f3a6fd7a8439f575d9ce17bae/src/python/tmclient/api.py#L1667
[b]: https://stackoverflow.com/a/22567429/459543

**Comment by dvischi** [comment]
Indeed, as discussed during our last meeting, the issue is related to the serialization.
As said, this is also the reason why we cannot simply send a byte stream.

I agree in the long-term approach. We should do so as soon as we refactor the API.

**Comment by riccardomurri** [comment]
Merged in branch "develop" as commit 3f20d92 (I had to change it to manually resolve a merge conflict).

**Pull request closed without merging**